hamilton/lifecycle/api.py

```diff
@@ -227,7 +254,96 @@ def post_node_execute(
         )
 
 
+class GraphExecutionHook(BasePreGraphExecute, BasePostGraphExecute):
+    """Implement this to execute code before and after graph execution. This is useful for logging, etc..."""
+
+    @override
+    @final
+    def post_graph_execute(
+        self,
+        *,
+        run_id: str,
+        graph: "FunctionGraph",
+        success: bool,
+        error: Optional[Exception],
+        results: Optional[Dict[str, Any]],
+    ):
+        return self.run_after_graph_execution(
+            graph=HamiltonGraph.from_graph(graph), success=success, error=error, results=results
+        )
+
+    @override
+    @final
+    def pre_graph_execute(
+        self,
+        *,
+        run_id: str,
+        graph: "FunctionGraph",
+        final_vars: List[str],
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+    ):
+        all_nodes, user_defined_nodes = graph.get_upstream_nodes(final_vars, inputs, overrides)
```

Comment:
Add comments/docstring on all these functions -- nice to have for internal documentation, even though it's pretty clear